### ISSUE TITLE
f16: the causal mask's -1e9 sentinel overflows the format — attention masking needs a precision-aware value

### ISSUE THREAD

**Issue by lukstafi** [ahrefs/ocannl#548]
Blocks `gpt2_mini` at f16 on every backend, alongside its sibling (the `-inf` cutoff rejection). Filed separately because they present identically in a log and need opposite fixes.

This one **is** a genuine overflow. The causal mask is filled with `-1e9`, which exceeds f16's 65504 max finite value outright — no guard policy is involved, the constant simply does not fit the format.

`-1e9` is a conventional "large negative" stand-in for `-inf` chosen so that `exp` underflows to zero in f32. At f16 it is not representable, and the value that *would* work — `-inf` — is currently rejected by the cutoff described in the sibling issue. So the two defects interlock: the representable sentinel is refused by policy, and the policy-passing sentinel is unrepresentable.

## What the fix has to decide

A **precision-aware sentinel**, rather than a hard-coded constant. Options, each with a different cost:

1. Emit `-inf` once the sibling issue exempts it — cleanest, and makes the mask precision-independent, but couples the two fixes.
2. Scale the sentinel to the storage precision (something like `-65504` at f16, `-1e9` at f32) — decoupled, but introduces a per-precision magic number and quietly changes softmax numerics at the boundary.
3. Express the mask as a boolean/`Where` selection rather than an additive sentinel, removing the magnitude question entirely — largest change, and probably the right long-term shape.

Worth deciding deliberately rather than patching the constant, since attention masking appears in `nn_blocks` and any model built on it inherits the choice.

Sibling: the `-inf` cutoff rejection. Related: gh-492, gh-538.

**Comment by lukstafi** [comment]
Fixed on branch `claude/dazzling-chatterjee-95e516`, alongside #547.

**Option 1**, chosen deliberately over the alternatives. `lib/nn_blocks.ml` gains

```ocaml
let default_mask_fill = Float.neg_infinity
```

used by `multi_head_attention` and overridable per call via `?mask_fill`, threaded through `decoder_only_block`, `decoder_only` and `transformer_decoder_block` so the escape hatch is reachable where padding masks actually get used.

## One correction to the option list

Option 3 — "express the mask as a boolean/`Where` selection rather than an additive sentinel" — is already the current shape. The code is `where mask scores !.fill`, not `scores + fill`; `Where` was never the missing piece. What a selection cannot avoid is *what value the false branch gets*, and that is the whole magnitude question. Removing it entirely would mean making `softmax` itself mask-aware (reduce max and sum over live positions only, never materializing a sentinel), which is a real design, but a considerably larger one than option 3 sounds — and it would have to reach `einmax1`'s reduction, not just the mask site.

Option 2 (per-precision magic number) I'd argue against on the issue's own terms: it makes every storage precision a place where softmax numerics can quietly differ, and it needs re-deciding for every format added (fp8 is already on the roadmap per gh-492).

## The cost of option 1, stated

The one thing `-1e9` bought was that a **fully** masked row (no live position at all) degrades to a uniform distribution rather than NaN — with an all-`-inf` row, the max-subtraction gives `-inf - -inf = nan`. Causal masks always leave the diagonal live and are unaffected, which is why nothing in-tree changes. A padding mask that can cover a whole row should pass a finite `?mask_fill` (e.g. `-1e4`, which still underflows `exp` to zero at every precision). That tradeoff is documented at the definition.

Coupling to #547 is real but was unavoidable in practice: since both defects gate the same lowering, `gpt2_mini` at f16 needs both regardless of the order they land in.

## Verification

`test/operations/half_softmax.ml` part 2 runs a causally masked softmax entirely at half — masked-out entries come out *exactly* zero (not merely small, which is what distinguishes a working sentinel from a saturated one), rows sum to 1, all entries finite. Byte-identical on cc and Metal.

`test/operations`, `test/einsum` and `test/ppx` are green with **no** promotion diffs: no existing golden moved, which is the expected result — `exp` underflows to exactly zero for both `-1e9` and `-inf` at f32, so this is numerics-preserving for every model that was already working.

**Comment by lukstafi** [comment]
Verified on hardware, on all four backends — closing.

`test/operations/half_softmax.ml` passes `dune runtest` against a single shared golden on cc and metal (M4 Max), cuda on rog-nv-wsl (RTX 5070 Ti), and hip on minix-amd-wsl (Radeon 8060S). The workload that motivated both issues, `gpt2_mini` at f16, now runs on each:

| backend | step time | step-0 loss |
|---|---|---|
| cc (M4 Max) | 6.2 s | 7.09813 |
| cuda (RTX 5070 Ti) | 209 ms | 7.09827 |
| hip (Radeon 8060S) | 53.6 ms | 7.09827 |

f32 reference is 7.09794, so every backend agrees to ~4e-4. gh-538's six blocked gpt2_mini f16 cells are unblocked.

**One thing the hardware run caught that local testing could not:** HIP's `exp` rounds the last mantissa bit differently from cc/metal/cuda — `2.1215e-1` where they give `2.1228e-1`, one ulp at 2^-13. Softmax outputs are `exp` results, so unlike the reduction identities no choice of inputs makes them exactly representable in half; the 4-significant-digit golden promoted locally would have failed on HIP. The test now prints above that divergence and puts the numeric content in a 5e-4 tolerance check against a double-precision reference. Worth knowing for any future half-precision golden.

Fixed in lukstafi/ocannl-staging#272.
